tests/test_sailor/test_sap_iot/test_wrappers.py

```diff
@@ -32,8 +44,58 @@ def simple_dataset(make_indicator, make_equipment_set):
-    assert aggregate_dataset._df.timestamp.min() == simple_dataset._df.timestamp.min()
+    assert aggregate_dataset._df.timestamp.min() <= simple_dataset._df.timestamp.min()
+    assert aggregate_dataset._df.timestamp.min() > simple_dataset._df.timestamp.min() - interval
     assert aggregate_dataset._df.timestamp.max() <= simple_dataset._df.timestamp.max()
+    assert aggregate_dataset._df.timestamp.max() > simple_dataset._df.timestamp.max() - interval
+
+
+def test_interpolation_happy_path(simple_dataset):
+    interval = pd.Timedelta('20min')
+    expected_count = math.ceil((simple_dataset.nominal_data_end - simple_dataset.nominal_data_start) / interval)
+
+    interpolated_dataset = simple_dataset.interpolate(interval)
+
+    assert all(interpolated_dataset._df.groupby('equipment_id').timestamp.count() == expected_count)
+    assert interpolated_dataset._indicator_set == simple_dataset._indicator_set
+    assert interpolated_dataset._df.timestamp.min() == simple_dataset.nominal_data_start
+    assert interpolated_dataset._df.timestamp.max() < simple_dataset.nominal_data_end
+
+
+@pytest.mark.parametrize('method,expect_ffill,expect_bfill', [
+    ('pad', True, False),
+    ('slinear', False, False),
+    ('bfill', False, True)
+])
+def test_interpolate_missing_data(simple_dataset, method, expect_ffill, expect_bfill):
+    test_dataset = simple_dataset.filter(['equipment_id_1'])
+    actual_start = test_dataset._df.timestamp.min()
+    actual_end = test_dataset._df.timestamp.max()
+
+    test_dataset.nominal_data_start = (actual_start - pd.Timedelta('1h')).round('1h')
+    test_dataset.nominal_data_end = actual_end + pd.Timedelta('1h')
+    remove_times = ((test_dataset._df.timestamp > actual_start + pd.Timedelta('1h')) &
+                    (test_dataset._df.timestamp < actual_start + pd.Timedelta('3h')))
+    for indicator in test_dataset._indicator_set:
+        test_dataset._df.loc[remove_times, indicator._unique_id] = float('NaN')
+
+    # this is not an regular test assert, but makes sure that during setup we actually introduced a 'hole' in the data
+    assert test_dataset._df.loc[remove_times].isnull().all().any()
+
+    interpolated_dataset = test_dataset.interpolate('1h', method=method)
+
+    assert interpolated_dataset._df.timestamp.min() == test_dataset.nominal_data_start
+    assert interpolated_dataset._df.timestamp.max() < test_dataset.nominal_data_end
+    remove_times = ((interpolated_dataset._df.timestamp > actual_start + pd.Timedelta('1h')) &
+                    (interpolated_dataset._df.timestamp < actual_start + pd.Timedelta('3h')))
+    assert not interpolated_dataset._df.loc[remove_times].isnull().any().any()
+
+    if expect_bfill:
+        assert not interpolated_dataset._df[interpolated_dataset._df.timestamp < actual_start].isnull().any().any()
+    if expect_ffill:
+        assert not interpolated_dataset._df[interpolated_dataset._df.timestamp > actual_end].isnull().any().any()
```

Comment:
I think here we're fine, since we set the nominal start/end date based on the actual start/end date of the data (lines 80/81). So there should always be data here that was interpolated if necessary.